sycl/doc/design/CommandGraph.md

```diff
@@ -1,66 +1,164 @@
-infrastructure.
+command-graph object: Explicit and Record & Replay API. Within the runtime they
+share a common infrastructure.
+
+## Nodes & Edges
+
+Nodes are represented by an internal `detail::node_impl` class which
+stores a `sycl::detail::CG` object for the command-group that the node
+represents. When a user adds a node to a graph using the explicit
+`command_graph<modifiable>::add()` API passing a CGF, in the implementation a
+`sycl::handler` object is created which finalizes the CGF, and after
+finalization the CG object from the handler is moved to the node.
+
+For creating a node in the graph using queue recording mode. When the
+`sycl::handler` from a queue submissions is finalized, if the queue the
+handler was created from is in the recording mode, then the command-group from
+finalization is added to the graph associated with the queue as a new node.
+
+Edges are stored in each node as lists of predecessor and successor nodes.
+
+## Sorting
+
+The current way graph nodes are linearized into execution order is using a
+reversed depth-first algorithm. Alternative algorithms, such as breadth-first,
+are possible and may give better performance on certain workloads/hardware. We
+are looking into giving the user control of this implementation detail.
+
+## Scheduler Integration
 
-## Scheduler integration
+When there are no requirements from accessors in a command-graph submission,
+the scheduler is bypassed and the underlying UR command-buffer is directly
+enqueued to a UR queue. If there are accessor requirements, the UR
+command-buffer for the executable graph needs to be enqueued by the scheduler.
 
-When there are no requirements for accessors in a command graph the scheduler
-is bypassed and it is directly enqueued to a command buffer. If 
-there are requirements, commands need to be enqueued by the scheduler. 
+When individual graph nodes have requirements from SYCL accessors, the
+underlying `sycl::detail::CG` object stored in the node is copied and passed to
+the scheduler for adding to the UR command-buffer, otherwise the node can
+be appended directly as a command in the UR command-buffer.
 
 ## Memory handling: Buffer and Accessor
 
 There is no extra support for Graph-specific USM allocations in the current
 proposal. Memory operations will be supported subsequently by the current
 implementation starting with `memcpy`.
 
-Buffers and accessors are supported in a command graph. Following restrictions
+Buffers and accessors are supported in a command-graph. Following restrictions
 are required to adapt buffers and their lifetime to a lazy work execution model:
 
 - Lifetime of a buffer with host data will be extended by copying the underlying
 data.
-- Host accessor on buffer that are used by a command graph are prohibited.
-- Copy-back behavior on destruction of a buffer is prohibited. 
+- Host accessor on buffer that are used by a command-graph are prohibited.
```

Comment:
```suggestion
- Host accessors on buffers that are currently used by a command-graph are prohibited.
```
I think the wording here was a bit ambiguous in that it could suggest that a buffer which had been used in a command-graph but that command-graph has since been destroyed would be prohibited, which I don't think we want.